core/2d/CCParticleSystem.cpp

```diff
@@ -634,6 +660,66 @@ void ParticleSystem::addParticles(int count)
         _particleData.posy[i] = _sourcePosition.y + _posVar.y * RANDOM_M11(&RANDSEED);
     }
 
+    if (animationCellIndex == -1 && _isEmitterAnimated)
+    {
+        for (int i = start; i < _particleCount; ++i)
+        {
+            int rand0                      = abs(RANDOM_M11(&RANDSEED) * _animIndexCount);
+            _particleData.animCellIndex[i] = MIN(rand0, _animIndexCount - 1);
+        }
+    }
+
+    if (animationCellIndex != -1)
+    {
+        for (int i = start; i < _particleCount; ++i)
+        {
+            _particleData.animCellIndex[i] = animationCellIndex;
+        }
+    }
+
+    if (animationIndex == -1 && !_animations.empty())
+    {
+        if (_randomAnimations.empty())
+            setMultiAnimationRandom();
+
+        for (int i = start; i < _particleCount; ++i)
+        {
+            int rand0 = abs(RANDOM_M11(&RANDSEED) * _randomAnimations.size());
+            int index = MIN(rand0, _randomAnimations.size() - 1);
+            _particleData.animIndex[i] = _randomAnimations[index];
+            auto& descriptor = _animations.at(_particleData.animIndex[i]);
+            _particleData.animTimeLength[i] =
+                descriptor.animationSpeed + descriptor.animationSpeedVariance * RANDOM_M11(&RANDSEED);
+        }
+    }
+
+    if (_isEmitterAnimated || _isLoopAnimated)
+    {
+        for (int i = start; i < _particleCount; ++i)
+        {
+            _particleData.animTimeDelta[i] = 0;
+        }
+    }
+
+    if (_isLoopAnimated && _animations.empty())
```

Comment:
Does the logic can be merge with above code since the behavior is same: `clear the _particleData.animTimeDelta to zero`

Reply:
Such as:
```cpp
if ((_isEmitterAnimated || _isLoopAnimated) ||  (_isLoopAnimated && _animations.empty()))
    std::fill_n( _particleData.animTimeDelta + start, _particleCount - start, 0);
```

Reply:
Aha: the condition expression: `(_isEmitterAnimated || _isLoopAnimated) ||  (_isLoopAnimated && _animations.empty())` equals to `(_isEmitterAnimated || _isLoopAnimated)`, so should be:
```cpp
if (_isEmitterAnimated || _isLoopAnimated)
    std::fill_n( _particleData.animTimeDelta + start, _particleCount - start, 0);
```

Reply:
And you can check other loop fill float array with 0, and all of them could be replaced with `std::fill_n`

Reply:
Actually that `(_isLoopAnimated && _animations.empty())` is not important because loop animations require an animation descriptor to work, so it doesn't hurt to leave that value uninitialized.